compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/parser/BLangNodeTransformer.java

```diff
@@ -5860,23 +5861,35 @@ private Object getIntegerLiteral(Node literal, String nodeValue, SyntaxKind sign
         return null;
     }
 
+    private void logParseLongError(Node literal, String originalNodeValue, SyntaxKind sign, DiagnosticCode code1,
+                                   DiagnosticCode code2) {
+        Location pos = getPosition(literal);
+        if (sign == SyntaxKind.MINUS_TOKEN) {
+            pos = new BLangDiagnosticLocation(pos.lineRange().filePath(),
+                    pos.lineRange().startLine().line(),
+                    pos.lineRange().endLine().line(),
+                    pos.lineRange().startLine().offset() - 1,
+                    pos.lineRange().endLine().offset());
+            dlog.error(pos, code1, originalNodeValue);
+        } else {
+            dlog.error(pos, code2, originalNodeValue);
+        }
+    }
+
     private Object parseLong(Node literal, String originalNodeValue,
                              String processedNodeValue, int radix, SyntaxKind sign,
                              DiagnosticCode code1, DiagnosticCode code2) {
+        // Since 9223372036854775808 is to overflow in Ballerina and -9223372036854775808 is a unary-expr where the
+        // expression is 9223372036854775808, -9223372036854775808 should also overflow even though it is within bounds
+        // for Java Long
+        if (originalNodeValue.equals(INT_MIN_OVERFLOW_VAL)) {
+            logParseLongError(literal, originalNodeValue, sign, code1, code2);
+            return originalNodeValue;
+        }
         try {
             return Long.parseLong(processedNodeValue, radix);
         } catch (Exception e) {
-            Location pos = getPosition(literal);
-            if (sign == SyntaxKind.MINUS_TOKEN) {
-                pos = new BLangDiagnosticLocation(pos.lineRange().filePath(),
-                                        pos.lineRange().startLine().line(),
-                                        pos.lineRange().endLine().line(),
-                                        pos.lineRange().startLine().offset() - 1,
-                                        pos.lineRange().endLine().offset());
-                dlog.error(pos, code1, originalNodeValue);
-            } else {
-                dlog.error(pos, code2, originalNodeValue);
-            }
+            logParseLongError(literal, originalNodeValue, sign, code1, code2);
         }
         return originalNodeValue;
     }
```

Comment:
This logic should move to `typechecker`. according to this(https://github.com/ballerina-platform/ballerina-spec/issues/943) `program 1` is a compile-time error and `program 2` should compile fine. 

Program 1
```ballerina
public function main() {
    int k = 999999999999999999999999999999;
}
```
Program 2
```ballerina
public function main() {
    float k = 999999999999999999999999999999;
}
```